jax/_src/cudnn/fused_attention_stablehlo.py

```diff
@@ -0,0 +1,728 @@
+  _check_valid_batch_dims(batch_dims)
+  query, key, value, bias, mask, activation, fwd_output, grad_output = batched_args
+  query_bdim = batch_dims[0]
+  out_bdims = query_bdim, query_bdim, query_bdim
+
+  *batch_tuple, q_seq_len, num_heads, head_dim = query.shape
+  *_, kv_seq_len, _, _ = key.shape
+  new_batch = reduce(operator.mul, batch_tuple)
+  has_bias, has_mask = variadic_args
+  # reshape to 4D shape
+  query = jnp.reshape(query, (new_batch, q_seq_len, num_heads, head_dim))
+  key = jnp.reshape(key, (new_batch, kv_seq_len, num_heads, head_dim))
+  value = jnp.reshape(value, (new_batch, kv_seq_len, num_heads, head_dim))
+  if has_bias:
+    bias = jnp.reshape(bias, (new_batch, num_heads, q_seq_len, kv_seq_len))
+  if has_mask:
+    mask = jnp.reshape(mask, (new_batch, num_heads, q_seq_len, kv_seq_len))
+  if is_flash_attention:
+    activation = jnp.reshape(activation, (new_batch, num_heads, q_seq_len))
+  else:
+    activation = jnp.reshape(activation, (new_batch, num_heads, q_seq_len, kv_seq_len))
+  fwd_output = jnp.reshape(fwd_output, (new_batch, q_seq_len, num_heads, head_dim))
+  grad_output = jnp.reshape(grad_output, (new_batch, q_seq_len, num_heads, head_dim))
+
+  grad_query, grad_key, grad_value = _dot_product_attention_bwd_p_wrapper.bind(
+    query, key, value, bias,
+    mask, activation, fwd_output, grad_output,
+    scale=scale, seed=seed, dropout_rate=dropout_rate,
+    variadic_args=variadic_args, is_flash_attention=is_flash_attention,
+    is_causal_mask=is_causal_mask)
+  
+  # reshape to original shape
+  grad_query = jnp.reshape(grad_query, (*batch_tuple, q_seq_len, num_heads, head_dim))
+  grad_key = jnp.reshape(grad_key, (*batch_tuple, kv_seq_len, num_heads, head_dim))
+  grad_value = jnp.reshape(grad_value, (*batch_tuple, kv_seq_len, num_heads, head_dim))
+  grads = (grad_query, grad_key, grad_value)
+  return grads, out_bdims
+
+# custom partitioning
+def _get_padded_spec(arg_info):
+  spec = None if arg_info.sharding is None else arg_info.sharding.spec
+  ndim = arg_info.ndim
+  if spec is None:
+    return (None,) * ndim
+  assert len(spec) <= ndim
+  return spec + (None,) * (ndim - len(spec))
+
+# fwd custom partition
+def _infer_fwd_output_sharding(mesh, arg_shapes):
+  # only sharding on batch and num_head dim is allowed
+  # (*batch, q_seq, num_head, head)
+  query_spec = _get_padded_spec(arg_shapes[0])
+  # (*batch, kv_seq, num_head, head)
+  key_spec = _get_padded_spec(arg_shapes[1])
+  value_spec = _get_padded_spec(arg_shapes[2])
+  if not query_spec == key_spec == value_spec:
+    raise ValueError("Query, key and value should have same sharding.")
+  seq_spec = query_spec[-3]
+  head_spec = query_spec[-1]
+  if seq_spec != None:
+    raise ValueError("Sharding on sequence dim is not allowed.")
+  if head_spec != None:
+    raise ValueError("Sharding on head dim is not allowed.")
```

Comment:
Normally, JAX/XLA silently force None sharding on dimensions it doesn't support.
Why do you raise an error here?

Reply:
I don't know what is the best practice here so I just raise an error for not supported sharding.

Reply:
Fine for me.